jax/_src/util.py

```diff
@@ -222,12 +222,28 @@ def wrapper(*args, **kwargs):
 CacheInfo = namedtuple("CacheInfo", ["hits", "misses", "maxsize", "currsize"])
 
 def weakref_lru_cache(call: Callable, maxsize=2048):
+  """
+  Least recently used cache decorator with weakref support.
+
+  The cache will take a weakref to the first argument of the wrapped function
+  and strong refs to all subsequent operations. In all other respects it should
+  behave similar to `functools.lru_cache`.
+  """
   cache: Dict[Any, Any] = {}
   hits = misses = 0
   lock = threading.Lock()
 
   def remove_key(tctx, args, kwargs, weak_arg):
-    del cache[(weak_arg, tctx, args, kwargs)]
+    k = (weak_arg, tctx, args, kwargs)
+    try:
+      # This has a chance to race with the iteration in next(iter(cache)),
+      # but we cannot lock because GC can get triggered synchronously inside
```

Comment:
You might be able to avoid gc if you use something like .pop() to extract the element inside the critical section but do not delete it until after? Just an idea, not mandatory.

Reply:
This code runs like an interrupt on a microprocessor whenever the GC runs. This means that we cannot acquire any locks. We either have to make the code 'lock-free' where the atomic unit of operation is a single python op (pop isn't available on dictionaries, but we could add some sort of pop_front c++ function?), move all this code to c++ where the assumptions are different, or have some sort of new lock type which runs these callbacks 'at the end of the next critical section'.

Reply:
I think callback should acquire `RLock` here, since [Google Styleguide says "Do not rely on the atomicity of built-in types."](https://google.github.io/styleguide/pyguide.html#218-threading)
It seems that `RLock` won't cause deadlock, see https://github.com/python/cpython/pull/5216.
Of course, `RLock` cannot fix the race condition(maybe "interleaving execution of weakref callback and next(iter(cache))" is more precise) here.

Reply:
In addition, it seems that there is another race condition without `RLock`:
1. lock was acquired in Thread 1, `del_k` was successfully computed.
2. `gc.collect` was triggered in Thread 2 and callback running in Thread 2 remove `del_k` from `cache`
3. `del cache[del_k]` in Thread 1 raise `KeyError`.

Reply:
To fix GC race condition, I think temporarily disable gc is better than catch runtime error.
```python
@contextmanager
def gc_disabled():
  import gc
  gc.disable()
  yield
  gc.enable()

with gc_disabled():
  del cache[next(iter(cache))]
```

Reply:
Disabling the gc doesn't prevent another python thread from dropping the final ref. It is also a global value which can race with other threads causing an indeterminate state for the gc. Libraries should not be randomly setting this value. SimpleQueue doesn't use RLock but instead is directly implemented via the python GIL.

This is probably the right fix for the race condition you mentioned (unless we want to move it to c so all the caching logic can be properly GIL protected and atomic):
```
try:
  del cache[next(iter(cache))]
except RuntimeError, KeyError:
  ...
```

Reply:
If another python thread from dropping the final ref, the callback should be invoked in another thread, and RLock would ensure `cache` "owned" by current thread current control flow.
By referencing cpython#5216, I mean reentrant can avoid deadlock when gc. And SQLAlchemy use RLock successfully https://github.com/python/cpython/issues/59181#issuecomment-1093581337.
Alternatively, we could do cleanup in another thread(using threadpool) and acquire nonreentrant lock.

Reply:
disabling gc won't cause intermediate state, at least if there is no gc callback that release gil, since gc is totally under gil. The only concern is that will set gc to enable regardless the old state of gc and we dont have a reliable way to save the old state.

Reply:
Not a big fan of the thread(pool?) idea either. While it is technically correct, it is very heavyweight and makes it so that the weakref callback can happen at some arbitrary time in the future.

The ideal would be LockWithWorkQueue (or a scoped protection against weakref callbacks), which would allow:
```
lock = LockWithWorkQueue()
with lock:
   # ... locks...
   
lock.call_locked(lambda: ...) # either runs immediately if the lock is available, or runs at the end of the current lock scope.
```

Without this, you still have the problem (with rlock) that the gc call can happen at any point. In which case, you might as well do what I've done and catch the potential errors that come up. Again, without moving code to c, it is hard to make anything perfect here.